### PR TITLE
implemented parser for iOS lockdown daemon log

### LINKED ISSUE #4067: Parser for lockdownd.log from iOS Sysdiagnose
Create a parser for the logs of the lockdown daemon on iOS. The  lockdownd.log file can be found in iOS sysdiagnose dump at:
`<sysdiagnose_DATE_OS_DEVICE_BUILD>/logs/MobileLockdown/lockdownd.log`

Based on research by Mattia Epifani, Adrian Leong and Heather Mahalik
https://www.rapidr.nl/wp-content/uploads/2019/05/iOS_Bug_Reporting_for_Forensic_Purposes_1.2.pdf

lockdownd is a the iOS daemon process in charge of providing iOS system information and access to services.
https://jon-gabilondo-angulo-7635.medium.com/understanding-usbmux-and-the-ios-lockdown-service-7f2a1dfd07ae

### PR DESCRIPTION
**Pull request by rick-slin**
## One line description of pull request
implemented parser for iOS lockdown daemon log

## Description:
implemented parser for iOS lockdown daemon log

**Related issue (if applicable):** fixes #4067

## Notes:
All contributions to Plaso undergo [code
review](https://github.com/log2timeline/l2tdocs/blob/main/process/Code%20review%20process.asciidoc). This makes sure
that the code has appropriate test coverage and conforms to the Plaso [style
guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
  - [ ] Automated checks (Travis, Codecov, Codefactor )pass
  - [ ] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [ ] Reviewer assigned

### CODE REVIEW

File: plaso/parsers/ios_lockdownd.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfVFS.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): tokens from a parsed log line.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+
```
Comment (jleaniz) [inline plaso/parsers/ios_lockdownd.py:82]: remove empty line

File: test_data/lockdownd.log [hunk 2]
```diff
@@ -0,0 +1,159 @@
+10/13/21 07:57:42.865446 pid=69 mglog: libMobileGestalt MGBasebandSupport.c:183: No IMEI in CT mobile equipment info dictionary
```
Comment (jleaniz) [inline test_data/lockdownd.log:1]: Could we please rename the test data file ios_lockdownd.log? This way it's clearer what it's for.

**Review by jleaniz — CHANGES_REQUESTED** [review]
Thanks for the parser! Left some minor comments for you to address.

File: plaso/parsers/ios_lockdownd.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  _BODY_END = pyparsing.StringEnd() | _TIMESTAMP
+
+  _BODY = pyparsing.SkipTo(_BODY_END).setResultsName('body')
+
+  _LINE_GRAMMAR = _TIMESTAMP + _PID + _BODY + pyparsing.ZeroOrMore(
+      pyparsing.lineEnd())
+
+  LINE_STRUCTURES = [('log_entry', _LINE_GRAMMAR)]
+
+  def ParseRecord(self, parser_mediator, key, structure: object):
```
Comment (joachimmetz) [inline plaso/parsers/ios_lockdownd.py:63]: pytype information not yet supported by this code base due to version differences in pytype definitions

File: plaso/parsers/ios_lockdownd.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      structure (pyparsing.ParseResults): tokens from a parsed log line.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+    if key != 'log_entry':
+      raise errors.ParseError(
+          'Unable to parse record, unknown structure: {0:s}'.format(key))
+
+    year = 2000 + self._GetValueFromStructure(structure, 'two_digit_year')
```
Comment (joachimmetz) [inline plaso/parsers/ios_lockdownd.py:82]: What if _GetValueFromStructure returns None ?

File: tests/parsers/ios_lockdownd.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        'timestamp': '2021-10-13 07:57:42.869324'}
+
+    self.CheckEventValues(storage_writer, events[6], expected_event_values)
+
+    expected_event_values = {
+        'body': ' spawn_xpc_service_block_invoke: description of xpc reply: '
+                '<dictionary: 0x2029c5070> { count = 1, transaction: 0, voucher'
+                ' = 0x0, contents =\n	"XPCErrorDescription" => <string: '
+                '0x2029c5230> { length = 22, contents = "Connection '
+                'interrupted" }\n}',
```
Comment (joachimmetz) [inline tests/parsers/ios_lockdownd.py:42]: remove `\n` from the value also leading space can be removed

**Comment by joachimmetz** [comment]
left some comments for educational purposes, made the changes already

**Pull request merged**